### ISSUE TITLE
Feature Request: CPU performance optimization with MKL-DNN

### ISSUE THREAD

**Issue by mingfeima** [pytorch/pytorch#4186]
Hi, our team works on DL frameworks performance optimization on CPU.
PyTorch CPU performance can be significantly improved with [MKL-DNN](https://github.com/01org/mkl-dnn), an open source deep neural network library on Intel CPU platform. Currently MKL-DNN supports Conv2d, Relu, BatchNorm, Pooling, Concat, etc. The CPU performance improvement will benefit cloud users as well as HPC researchers who have access to Xeon clusters.
Will PyTorch be interested in the part?

**Comment by soumith** [comment]
hi @mingfeima . We are very interested in the settings where BCHW memory layout is preserved and weight layout is compatible with PyTorch's weight layout. Is MKLDNN shipped on Anaconda yet? (if not we can maybe static link)

**Comment by mingfeima** [comment]
Hi, @soumith . MKLDNN does have an internal memory layout, but the intention is to optimize convolution performance by reorder memory in a cache friendly manner, rather than  break down the native layout of a framework. Ground rule is that whenever you want to read a tensor outside  MKLDNN (e.g. dump the weight), it is in BCHW layout.

So clear boundaries of MKLDNN supported layer (e.g. conv2d) and unsupported layer should be defined and MKLDNN internal format only flow inside MKLDNN supported layers. If you treat CPU as an computation device as GPU, the data flow is actually similar. GPU needs to do memcpy from host/device, CPU needs to do memory reorder inside/outside MKLDNN.

Feedback from recent past NIPS indicates that there might be concerns that MKLDNN internal layout is opaque. The internal layout is not opaque actually, it represents a way of blocking the memory. Taking the commonly used [nChw16c](https://github.com/01org/mkl-dnn/blob/master/include/mkldnn.hpp#L565), 16c indicates 16 floats along the channel dim for vectorization purpose (512-bit SIMD), C = number of channels / 16, nChw dims are used for blocking of OpenMP parallel. You can also directly read an internal layout memory based on BCHW index. MKLDNN will select the best fit layout based on input tensor size, algorithms(direct, winograd, etc.) and hardware configuration (256-bit or 512-bit) to assure the best performance on a given CPU.

MKLDNN full package is not available on Anaconda yet at the moment and there is no recent plan.

**Comment by mingfeima** [comment]
Hi, @soumith . Still looking forward to your reply. I already uploaded [MKLDNN](https://anaconda.org/mingfeima/mkldnn) to Anaconda, you can give it a try.

**Comment by soumith** [comment]
I think we can benefit from MKLDNN, particularly on standard neural network workloads, but particularly on tensor operations like transpose-copy, etc.

At the moment, @cpuhrsch is making a list of areas in pytorch that need CPU improvements and prioritizing them. He shall share them with you and whichever areas are higher-priority and need MKL-DNN integration, we can get those going.

We still have reservations on using MKLDNN's special layout because that would introduce the need to carry opaque state around. But MKLDNN has some APIs that dont need using different layout.
In the meanwhile we will think on how to introduce different layouts into pytorch (and if it makes sense)

**Comment by mingfeima** [comment]
@soumith @cpuhrsch Glad to know! We can have more discussion when the list comes out.
Our target is to promote PyTorch CPU performance, besides MKLDNN integration we also work on other areas, such as fused RNN kernel on CPU, etc. We can re-prioritize our work according to the list.
@MlWoo has already submitted PR on element wise optimization #2764 and 3d convolution #4824.

**Comment by cpuhrsch** [comment]
Hello @mingfeima,

There is one workload ([Word Language Model](https://github.com/yf225/examples/tree/benchmark_test/word_language_model)) we've been working on recently where a) an efficient CPU implementation [does not exist](https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/FusedRNNKernel.c) b) it represents a [good proxy for other applications](https://gist.githubusercontent.com/yf225/40c0adb8bfb2a7b774fa266fb4bc409e/raw/20c67ebadbd75f41c6c9fd2e00b4b2562b60700a/mini_sequence_labeler.py) and c) pytorch has recently invested work in [optimizing](https://github.com/pytorch/pytorch/commit/7bd2db997e95bf69e67bee3be5b0ef73ef39a516#diff-63ea4b4ea36b3763ca8ed2806100426a) the [GPU version](https://github.com/pytorch/pytorch/commit/7bd2db997e95bf69e67bee3be5b0ef73ef39a516). This is a good starting point and fused RNN kernels can significantly speed up the code here. Let me know if this works for you.

Thanks,
Christian

**Comment by mingfeima** [comment]
@cpuhrsch sure, this definitely works! Actually we have finished most of coding for fusedRNN kernel for CPU, this one is on our list. Once the tuning and testing is done, we will propose a PR.
We have been using [deepspeech](https://github.com/SeanNaren/deepspeech.pytorch) as the benchmark workload, we will also take your workloads [(Word Language Model)](https://github.com/yf225/examples/tree/benchmark_test/word_language_model) and [sequence labeler](https://gist.githubusercontent.com/yf225/40c0adb8bfb2a7b774fa266fb4bc409e/raw/20c67ebadbd75f41c6c9fd2e00b4b2562b60700a/mini_sequence_labeler.py) for benchmark. 
Do you have any publicly available dataset for mini_sequence_labeler, or the input size for this workload? It would be easier for performance tuning if we knew the input size range.

**Comment by yf225** [comment]
Hi @mingfeima, for mini_sequence_labeler benchmark we currently use the script and dataset from here: https://github.com/yf225/benchmark/tree/perf_test_scripts/scripts/mini_sequence_labeler. Please feel free to use it for performance tuning as well.

**Comment by andrei-pokrovsky** [comment]
Hello! Just wondering, is there a branch of pytorch somewhere with MKL optimizations already integrated in ptic for convolutions or is this still work in progress? Thank you!

**Comment by soumith** [comment]
@andrei-pokrovsky we haven't yet integrated MKL-DNN integrations. However, we integrated NNPack convolutions. https://github.com/Maratyszcza/NNPACK

**Comment by mingfeima** [comment]
@andrei-pokrovsky, just coming back from Chinese New Year leave. We have a [repo](https://github.com/intel/pytorch) with MKL convolution optimization integrated and functionally verified. This repo is based pytorch v0.2 and we have been shifting to latest code base in our internal branch. This repo serves as a prototype design with CPU perf. optimization, more work is still WIP. We will propose PR gradually, following guidelines from PyTorch community of course.

**Comment by mingfeima** [comment]
@soumith @cpuhrsch @yf225 we have been looking into [word_language_model](https://github.com/yf225/examples/tree/benchmark_test/word_language_model) lately.
The low CPU performance is caused by several reasons, missing FusedRNNKernel is one of them but not the major one. After implementing forward and backward path of GRU and LSTM at aten/src/THNN/generic/FusedRNNKernel.c, the CPU performance gets only 10% performance boost on our Xeon Skylake 8180 server (2 sockets, 28 core per socket), which is quite abnormal.

1. the biggest issue we found is that the forward path and backward path are using two sets of OpenMP threads separately at runtime. The usual BKM when running a server CPU is to set OMP_NUM_THREADS = number of physical cores. And we observed from vtune that forward path is using a set of 56 OpenMP threads and backward path is using another 56. Once set OMP_NUM_THREADS to be 28, i got much better performance even losing 50% of the total computing power, since fwd/bwd path threads are executing peacefully. :(
The issue slows down not only RNN but also Convolution, and almost every CPU operator that depends on OpenMP for parallelization.
2. Besides GRU/LSTM layers, some other operators also contribute a lot to the total runtime, for instance:
- logsoftmax
- dropout
- norm
We will have these problems solved in an appropriate manner, step by step.
The analysis of mini sequence labeler just began, i will have you updated later.

**Comment by fmassa** [comment]
@mingfeima Thanks for the information!
I believe we have recently optimized `norm` in https://github.com/pytorch/pytorch/pull/5722
About the slowdowns due to double the number of threads than required, I thought that the backward pass of autograd only spawned threads if we were performing operations on GPUs? @colesbury 
But I might be missing something...

**Comment by apaszke** [comment]
@fmassa It always spawns the threads, but I think it's not so much about them (they will be idle if you run on CPU only), but about OMP having separate thread pools for forward and backward threads.

@mingfeima thanks for the update. A few notes from our side that might be relevant:

I looked into `logsoftmax` a few months back and concluded that it's heavily bottlenecked by functions like `exp` and `log` that prevent vectorization (I know `icc` can do it properly, but that doesn't solve it for us). Since then we've merged vectorized implementations of those functions into ATen or TH (can't remember now), so it should be relatively easy to get an order of magnitude improvement there.

We also added support for generating fused CPU kernels on the fly to the PyTorch JIT, but didn't proceed with the work because we haven't seen any real benefits from these fusions just yet. Are there really cases where you can get a 50% speedup?

**Comment by mingfeima** [comment]
@gautamkmr first of all, please be aware that MKL and MKL-DNN are two separate packages.
MKL contains general BLAS operation such as GEMM (matrix multiplication).
MKL-DNN contains Deep Neural Network operations such as convolution, etc.

Current conda package contains **MKL** but not **MKL-DNN**.
You need to compile from source to get MKL-DNN, install [MKLDNN](https://anaconda.org/mingfeima/mkldnn) and follow normal installation process will be enough.
Also we wrote [bkm](https://github.com/intel/pytorch#bkm-on-xeon) for cpu performance.

**Comment by cpuhrsch** [comment]
I'm looking into speeding up softmax and logsoftmax now that we have vectorized implementations of exp and log. However, I'm not sure if I should continue with this? @mingfeima is there a timeline on when you will release logsoftmax? Will this include a backward pass as well?

**Comment by yf225** [comment]
@mingfeima Also we noticed that MKL-DNN uses OpenMP 4.0 SIMD extensions which is not supported by MSVC and hence can't be used on Windows (it might be difficult to ask users to use ICC since it's proprietary). Do we have a solution in mind for Windows in this case?

**Comment by mingfeima** [comment]
@cpuhrsch we only PR mkldnn convolution at the moment but actually code for lot of other operators are already in our internal branch.

The real problem is that mkldnn utilizes unique data layouts for optimization purpose, what we do for the convolution layer at now is 1) reorder from bchw to internal 2) do computation 3) reorder from internal to bchw. If we reorder everytime for every operator, only convolution is beneficial using mkldnn.

So from the performance point of view, it is needed to cache the memory of mkldnn layout. Do reorder only once when entering and exiting for a consecutive mkldnn calls. This is what has been done on [caffe2/ideep](https://github.com/pytorch/pytorch/tree/master/caffe2/ideep). (ideep is a wrapper of mkldnn). However this is against @soumith initial proposal of reserving BCHW layout. So i can't simply follow what has been done on caffe2 or tensorflow...

I'm still investigating methods for caching the mkldnn memory, we have several plans but haven't got a perfect one.

Anyway, by vectoring exp and log is good enough at the stage, won't have serious performance gap with mkldnn solutions. Once we reach an agreement of caching mkldnn memory in ATen, both fwd/bwd of softmax will be provided.

**Comment by mingfeima** [comment]
@yf225 [mkldnn](https://github.com/intel/mkl-dnn#system-requirements) has Windows support, you can use MSVS compiler, icc is not needed there.

i disabled Windows support at the initial commit just because i haven't got the bandwidth to test it on Windows.

So what is the expectation of Windows support? Make it functional will be enough or make it as performant as Linux?

**Comment by yf225** [comment]
@mingfeima I guess the code will run slower on Windows since MSVC doesn't support OpenMP 4.0 things like SIMP? Alternatively we are experimenting with using ICC to compile, and see how much benefit it brings us in terms of performance on Windows.

**Comment by mratsim** [comment]
@mingfeima @apaszke Here is my analysis on softmax/logsoftmax optimization. Taken from [here](https://github.com/pytorch/pytorch/blob/a015d579ddce90831bb945bbefdd1aa1a3633f63/aten/src/ATen/native/SoftMax.cpp#L25-L77).

```C++
template<typename scalar_t, bool LogSoftMax>//, template<typename> class Epilogue>
void host_softmax(Tensor output, const Tensor & input_, const int64_t dim_){
  auto input = input_.contiguous();
  int64_t dim = maybe_wrap_dim(dim_, input.dim());
  if (input.dim() == 0) input = input.view(1);
  AT_CHECK(dim >=0 && dim < input.dim(), "dim must be non-negative and less than input dimensions");
  uint64_t outer_size = 1;
  uint64_t dim_size = input.size(dim);
  uint64_t inner_size = 1;
  for (int64_t i = 0; i < dim; ++i)
    outer_size *= input.size(i);
  for (int64_t i = dim + 1; i < input.dim(); ++i)
    inner_size *= input.size(i);
  uint64_t dim_stride = inner_size;
  uint64_t outer_stride = dim_size * dim_stride; 
  scalar_t *input_data_base  = input.data<scalar_t>();
  scalar_t *output_data_base = output.data<scalar_t>();
  LOG_SOFTMAX_SIZE_TYPE i, d;
#pragma omp parallel for private(i, d)
  for (i = 0; i < LOG_SOFTMAX_CAST_TYPE (outer_size * inner_size); i++)
  {
    uint64_t outer_idx = i / inner_size;  // costly division
    uint64_t inner_idx = i % inner_size;
    scalar_t *input_data  = input_data_base  + outer_idx * outer_stride + inner_idx;
    scalar_t *output_data = output_data_base + outer_idx * outer_stride + inner_idx;

    scalar_t max_input = -std::numeric_limits<scalar_t>::max();
    for (d = 0; d < LOG_SOFTMAX_CAST_TYPE dim_size; d++)       // 1st loop over the data
      max_input = std::max(max_input, input_data[d * dim_stride]);

    using accscalar_t =  acc_type<scalar_t, false>; 
    accscalar_t tmpsum = 0;
    for (d = 0; d < LOG_SOFTMAX_CAST_TYPE dim_size; d++){    // 2nd loop over the data
      scalar_t z = std::exp(input_data[d * dim_stride] - max_input);
      tmpsum += z;
      if (!LogSoftMax){ // branching in a very tight loop
        output_data[d*dim_stride] = z;
      }
    }
  
    if (LogSoftMax) // branching in a tight loop
       tmpsum = max_input + std::log(tmpsum);
    else
       tmpsum = 1 / tmpsum;

    for (d = 0; d < LOG_SOFTMAX_CAST_TYPE dim_size; d++)  // 3rd loop over the data
      if (LogSoftMax) // branching in a very tight loop
         output_data[d * dim_stride] = input_data[d * dim_stride] - tmpsum;
      else
         output_data[d * dim_stride] *= tmpsum; 
  }

}
```

In the main OpenMP loop:
  - you have 3 loops over the data
  - branching in tight loops
  - a division/modulo which are quite expensive.

Since logsoftmax/softmax is known at compile-time that branching can be templatized.
Also you can compute the max and sum of exp in a streaming fashion (one pass similar to streaming mean/variance) [as proposed in this blog post](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html). This is especially beneficial for language modelling where vocabulary size is huge.

In my own tensor library, using NCHW layout, this is the resulting code, it is at least 25% faster than 3 for loops. Code is in Nim (a compiled-to-C/C++ and Javascript language with the syntax similar to Python).

```Nim
proc streaming_max_sumexp*[T](t: Tensor[T]): tuple[max:T, sumexp: T] {.noSideEffect, inline.}=
  result.max = -Inf.T   # will store the streaming max of the tensor
  result.sumexp = 0.T   # will store the streaming sum of exp of the tensor

  for x in t:
    if x <= result.max:
      result.sumexp += exp(x - result.max)
    else:
      result.sumexp = result.sumexp * exp(result.max - x) + 1
      result.max = x

proc stable_softmax*[T](x, max, sumexp: T): T {.noSideEffect, inline.}=
  # Numerically stable streaming softmax helper
  result = exp(x - max) / sumexp

proc softmax*[T](input: Tensor[T]): Tensor[T] {.noInit.} =
  ## For each sample in a tensor:
  ##   do an exponential normalization of each of its class features xi
  ##   ``exp(xi) / ∑i exp(xi)``
  ##
  ## Input:
  ##   - A tensor of shape [batch_size, number_of_classes]
  ## Output:
  ##   - A tensor of shape [batch_size, number_of_classes]

  let batch_size = input.shape[0]
  result = zeros_like(input)

  for i in 0||(batch_size-1): # || is a OpenMP parallel loop
    let (max, sumexp) = input[i, _].streaming_max_sumexp

    var res_slice = result[i, _]

    apply2_inline(res_slice, input[i, _]):  # This is an inline Nim template for parallel tensor iteration
      stable_softmax(y, max, sumexp) # y refers to the second argument of the template
```

Reference:
  - [Softmax](https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/nn_primitives/nnp_softmax.nim#L21-L40)
  - [Streaming_max_sumexp](https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/nn_primitives/private/p_logsumexp.nim#L30-L39)
  - [Stable softmax helper](https://github.com/mratsim/Arraymancer/blob/249218eb64476c5aec34ca15d0dcf0955fbb418c/src/nn_primitives/private/p_logsumexp.nim#L57-L59)

Regarding logsoftmax, a fused softmax_cross_entropy similar to Tensorflow's would be much faster than cross_entropy(logsoftmax(x)) as you can do everything in fewer loops loops.

**Comment by fmassa** [comment]
Hi @mratsim,
A comment on your analysis: I believe there is no branching in tight loops, because `LogSoftMax` is a template parameter and gets compiled away from the code (as you mention already).
Also, there is some improvements on both `softmax` and `logsoftmax` that are being implemented in https://github.com/pytorch/pytorch/pull/7375.

But, I agree that we could implement a dedicated function for `cross_entropy`, which currently dispatches to `nll_loss` + `logsoftmax`.

**Comment by MlWoo** [comment]
@cpuhrsch  If I do not miss something, the vectorization optimization is applied to softmax and logsoftmax with the last deimension on a contiguous tensor. At first, I have to admit that calculating softmax and logsoftmax with the last deimension on a contiguous tensor is frequent in classifier. So the PR LGTM. 

However, there are some questions to be confirmed.
1.  As a benchmark, perf on non-contiguous tensor maybe needs to be provided. The code shows that softmax clones the tensor to make a new contiguous one first if handling with a non-contigous tensor.  I doubt the benefit from vectorization could not cover overhead of copy in some case, not to mention comparison with streaming the operation( the method provided by @mratsim).  
2. If the dimension is not the last one, there are some problems which have been pointed out by @mratsim. 

So there are some room to be better on these operations. But opt further on softmax is not urgent in consideration of the practice of classifier.

**Comment by cpuhrsch** [comment]
@MlWoo Thanks for the review! Yes, indeed it's only over the last dimension and contiguous Tensors only. I'll look more closely into what you and @mratsim mentioned.
